packages/angular/cli/src/commands/generate/cli.ts

```diff
@@ -162,4 +160,45 @@ export class GenerateCommandModule
 
     return `${commandName}${positionalArgs ? ' ' + positionalArgs : ''}`;
   }
+
+  private async getSchematicsToRegister(): Promise<
+    [schematicName: string, collectionName: string][]
+  > {
+    const [, schematicNameFromArgs] = this.parseSchematicInfo(
+      // positional = [generate, component] or [generate]
+      this.context.args.positional[1],
+    );
+
+    /**
+     * - *key*: schematic name
+     * - *value*: collection name
+     */
+    const schematicsToRegister = new Map<string, string>();
+    const collectionNames = await this.getCollectionNames();
+
+    for (const collectionName of collectionNames) {
+      const workflow = this.getOrCreateWorkflowForBuilder(collectionName);
+      const collection = workflow.engine.createCollection(collectionName);
+      const schematicsInCollection = collection.description.schematics;
+
+      // We cannot use `collection.listSchematicNames()` as this doesn't return hidden schematics.
+      for (const schematicName of Object.keys(schematicsInCollection)) {
+        if (schematicsInCollection[schematicName].private) {
+          continue;
+        }
+
+        if (schematicNameFromArgs === schematicName) {
+          // No need to process all schematics since we know which one the user invoked.
+          return [[schematicName, collectionName]];
+        }
```

Comment:
Nit: This feels like it's breaking abstraction to me. The point of this function is to get all the schematics, right? Not to find a particular schematic. I think it would be up to the caller to look for a particular schematic. If you want to short-circuit the logic, this could be rewritten as a generator, and then the caller can just `break;` when it finds what it wants. The `.sort()` call might make that impractical though.

Reply:
The point of this method is actually get the schematics that needs to be registered, hence the name `getSchematicsToRegister`. 

IIf `schematicNameFromArgs` matches one of the schematic we only need to register the single schematic and ignore all the others, while when `schematicNameFromArgs` doesn't match any known schematic we need to register all schematics. 

I don't think a generator is the best approach here, since what we want is to get a list of schematics that we should register. 

But I did refactored this a bit more.

Reply:
I would agree about the abstraction that it should return the full list of schematics and the caller should filter. But in this case then the method should renamed to something else like `getListOfSchematics` etc.. which IMHO in this case isn't really needed for 2 points, this is a private method, which in case in future we need to get a full list of schematics we can always update the code later on.

That said, if you feel we should go with your approach I don't have not totally opposed to the idea.

Reply:
I guess `getListOfSchematics()` and registering all of them until we find the one we want makes more sense to me than `getSchematicsToRegister()` and doing all unless we find the one we want early.

I'm happy to defer to you here, it just stood out to me. I don't think it's a big enough deal to be a blocker for this PR.

Reply:
I added both method. The reason I kept `getSchematicsToRegister` is to reduce the size of the builder method and make it readable.

Reply:
I like your approach here of making `getSchematicsToRegister()` wrap `getSchematics()` and adding the early return there.